### PR TITLE
[ci] Track active_parties in scan aggregation for better performance.

### LINKED ISSUE #732: Scan aggregator makes a sequential scan over round_party_totals
[ This issue was auto-migrated from DA's internal repo (DACH-NY/canton-network-node#14176). Original author: @moritzkiefer-da ]

This is clearly not great since that table will grow forever.

image

Have not looked at query plans or whatever.

### PR DESCRIPTION
**Pull request by ray-roestenburg-da**
Fixes #732 

- [x] Add a test with an empty `active_parties` table and a non-empty `round_party_totals`, verifying that  `initiate_active_parties` works an subsequent total and cumulative are correct.
- [x] Add a test where not all parties are active in all rounds. 

#### Cluster Testing
- [ ] If a cluster test is required, comment `/cluster_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a hard-migration test is required (from the latest release), comment `/hdm_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.

#### PR Guidelines
- [ ] Include any change that might be observable by our partners or affect their deployment in the [release notes](https://github.com/DACH-NY/canton-network-node/blob/main/cluster/images/docs/src/release_notes.rst).
- [ ] Specify fixed issues with `Fixes #n`, and mention issues worked on using `#n`
- [ ] Include a screenshot for frontend-related PRs - see [README](https://github.com/DACH-NY/canton-network-node#running-and-debugging-integration-tests) or use your favorite screenshot tool

#### Merge Guidelines
- [ ] Make the git commit message look sensible when squash-merging on GitHub (most likely: just copy your PR description).

### CODE REVIEW

File: apps/common/src/main/resources/db/migration/canton-network/postgres/stable/V040__active_parties.sql [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+);
+
+create index active_parties_sid_p_cr_idx
+    on active_parties (store_id, party, closed_round);
+
+create or replace procedure initialize_active_parties()
+language plpgsql
+as $$
+begin
+    if not exists (select 1 from active_parties limit 1) then
```
Comment (ray-roestenburg-da) [inline apps/common/src/main/resources/db/migration/canton-network/postgres/stable/V040__active_parties.sql:16]: This will show as seq scan but is one row only since it stops immediately so no concern.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala [hunk 2]
```diff
@@ -716,18 +716,28 @@ final class ScanAggregator(
...
-       * and that it is safe to start the aggregation from any previously aggregated round, even if some parties are missing in some rounds.
-       * finally inserting the sums and cumulative sums in the round_party_totals table.
-       */
-      sql"""
+
+    /** Initialization of active_parties is done in an idempotent stored proc (Effect.Write).
+      * This is done outside of flyway migration so that it does not impact scan app startup time.
+      */
+    def initializeActiveParties(): DBIOAction[Unit, NoStream, Effect.Write] = {
+      sqlu"""call initialize_active_parties()"""
```
Comment (moritzkiefer-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:724]: how fast is it on CILR scale data?

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala [hunk 3]
```diff
@@ -818,39 +828,66 @@ final class ScanAggregator(
...
+        select  rpt.party as party,
+                coalesce(cumulative_app_rewards, 0) as prev_cumulative_app_rewards,
+                coalesce(cumulative_validator_rewards, 0) as prev_cumulative_validator_rewards,
+                coalesce(cumulative_traffic_purchased, 0) as prev_cumulative_traffic_purchased,
+                coalesce(cumulative_traffic_purchased_cc_spent, 0) as prev_cumulative_traffic_purchased_cc_spent,
+                coalesce(cumulative_traffic_num_purchases, 0) as prev_cumulative_traffic_num_purchases,
+                coalesce(cumulative_change_to_initial_amount_as_of_round_zero, 0) as prev_cumulative_change_to_initial_amount_as_of_round_zero,
+                coalesce(cumulative_change_to_holding_fees_rate, 0) as prev_cumulative_change_to_holding_fees_rate
+        from    round_party_totals rpt
+        join    active_parties ap
```
Comment (moritzkiefer-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:841]: I have a hard time understanding why this join is correct. What about this scenario:
1. Party is active in round 1, it has a total of app rewards (or whatever else, doesn't really matter) of 42. This means we have an entry for that party in round 1 in `active_parties`.
2. Party is not active in round 2. This means we have no entry for that party in round 2 in `active_parties`
3. Party is active again in round 3.
4. Now you aggregate from round 2 to round 3. Doesn't this mean that the join here will not include any entry for that party so you consider the previous total 0 instead of 42?

Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:841]: Rows in active_parties are mostly updated, unless new parties are active that have never been seen in the txlog.
There is only one row per party in the table (not considering the store_id)
1. yes. 
2. yes, but the `round_party_totals` also does not have it (not aggregated yet), it has an entry for that party for round 1 (and maybe one for round 0, who knows), and this is joined, and seen as the state before aggregation.
3. the active_parties only has one entry per party. so you can't be in the state where there is a round 1 and round 3 here (the most recent round it was 'active' in). If the party was active in round 3, before we aggregate, active_parties is already reflecting that, (and so is round_party_totals as part of the previous aggregation run) or it will only update that the next time as previous totals.
4. Depending on the data, aggregation is done for 0 .. many rounds at a time, we do a running total 'sum partition over' for the data for the new rounds, compared to the totals for the last known round per party
5. the aggregation is not per round, but for all rounds between previously aggregated, and what is read from tx log, up to 'last closed round'

Comment (moritzkiefer-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:841]: Let's say aggregation happens each round so yes I'm aware we can skip over rounds but if we're fast enough we don't.

In the example I mentioned above, what is the state of `active_parties` after aggregation of round 2? Does the party still have an entry with round 1? Do we update it to round 2 even though it has not been active in round 2? If the latter, I think I don't understand what "active" means.

Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:841]: > In the example I mentioned above, what is the state of active_parties after aggregation of round 2? Does the party still have an entry with round 1?

party has entry with round 1 in active_parties.

> Do we update it to round 2 even though it has not been active in round 2? 

no

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala (hunk 2 above)
Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:724]: On a scan clone from CILR it completed in  ~48 seconds.

Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:724]: (when it needed to initialize of course, once initialized the time it takes is negligible)

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala (hunk 3 above)
Comment (moritzkiefer-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:841]: ok I think I get it now so previous_totals is not the totals of the round we're aggregating from, it's the previous per partyt which can be different per party.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala [hunk 4]
```diff
@@ -818,39 +828,66 @@ final class ScanAggregator(
...
+                sum(nt.traffic_purchased) over (partition by nt.party order by round) + coalesce(pt.prev_cumulative_traffic_purchased,0) as cumulative_traffic_purchased,
+                sum(nt.traffic_purchased_cc_spent) over (partition by nt.party order by round) + coalesce(pt.prev_cumulative_traffic_purchased_cc_spent,0) as cumulative_traffic_purchased_cc_spent,
+                sum(nt.traffic_num_purchases) over (partition by nt.party order by round) + coalesce(pt.prev_cumulative_traffic_num_purchases,0) as cumulative_traffic_num_purchases,
+                sum(nt.change_to_initial_amount_as_of_round_zero) over (partition by nt.party order by round) + coalesce(pt.prev_cumulative_change_to_initial_amount_as_of_round_zero,0) as cumulative_change_to_initial_amount_as_of_round_zero,
+                sum(nt.change_to_holding_fees_rate) over (partition by nt.party order by round) + coalesce(pt.prev_cumulative_change_to_holding_fees_rate,0) as cumulative_change_to_holding_fees_rate
+      from      new_totals nt
+      left join previous_totals pt
+      on        nt.party = pt.party;
+
+      insert into active_parties (
```
Comment (moritzkiefer-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala]: We have an insert here and then an update below. Why do we need both? Aren't you just doing an upsert i.e. insert the max round for each party from temp_cumulative_totals and `on conflict do update round = max_round`?

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala [hunk 5]
```diff
@@ -390,126 +390,112 @@ class ScanAggregatorTest
...
+            store,
+            lastRound,
+            lastRound.toInt,
+          )
+          assertActiveParties(store, nrParties, lastRound)
+        }
+      }
+    }
+
+    "Aggregate round_party_totals where some parties are not always taking part" in {
```
Comment (moritzkiefer-da) [inline apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala]:
```suggestion
    "Aggregate round_party_totals where some parties are not active in all rounds" in {
```

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala [hunk 6]
```diff
@@ -390,126 +390,112 @@ class ScanAggregatorTest
...
+      cleanup()
+      val (aggr, store) = mkAggregator(dsoParty).futureValue
+      val lastRound = 5L
+      val firstParties = Seq(1, 2)
+      val firstRounds = Seq(0L, 1L, 2L)
+      val lastParties = Seq(3, 4, 5)
+      val lastRounds = Seq(3L, 4L, 5L)
+      for {
+        first <- generateRoundPartyTotalsRange(store, firstRounds, firstParties)
+        last <- generateRoundPartyTotalsRange(store, lastRounds, lastParties)
```
Comment (moritzkiefer-da) [inline apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala:473]: can we make sure to cover the scenario where a party is not active for some rounds but is then active again later? afaik you only cover the scenario here where they are either not at all active at the start or stop activity completely after firstRounds but not where there are gaps

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala (hunk 2 above)
Comment (moritzkiefer-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala:724]: sounds good, thanks

**Review by moritzkiefer-da — APPROVED** [review]
Thanks! I think I convinced myself now that it works but the tests don't seem as detailed as they should be imho

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala (hunk 6 above)
Comment (ray-roestenburg-da) [reply apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala:473]: Sounds good, will do.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala (hunk 4 above)
Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala]: I read around and convinced myself that it was faster to insert select only what is not there. Most of the time it's updates and conflicting on insert most of the time seemed like a bad idea. But on another note I see now that I have a strange leftover here! this insert should not be there, only the next one. Will fix.

Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala]: But maybe I am wrong about this assumption and upserts might be better, I'll check this again, thanks for spotting this.

Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala]: It might be better to do an upsert, postgres is supposedly fast at it, I was probably overworried about conflicting on inserts a lot.. I'll change it.

Comment (ray-roestenburg-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/ScanAggregator.scala]: Using upsert now.

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala (hunk 6 above)
Comment (ray-roestenburg-da) [reply apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/store/db/ScanAggregatorTest.scala:473]: @moritzkiefer-da Added a test for this scenario.
